### DISCUSSION TITLE
First playbook, first problem

### DISCUSSION THREAD

**Discussion by gpow81** [ansible-collections/ibm_zos_core#380] [Q&A] [answered]
Hello, I hope I am asking this in the right place. I will be working on a project which will involve running ansible playbooks on the z/OS. I wanted to start learning with something simple and I think I am close but I stuck on a python syntax error.  

>>>SyntaxError: Non-UTF-8 code starting with '\xa4' in file <stdin> on line 2, but no encoding declared; see https://python.org/dev/peps/pep-0263/ for details

On the z/OS I have
- python 3.10.5
- zoau-1.2.0.1.
- USS is working
- group_vars with all.yml from one of the samples and host_vars.yml with paths. 
 
On my workstation I have ansible with installed ibm_zos_core collection. 
Whichever module I would try to run (or even just -m ping) , I am getting same error about the encoding. Funny thing is that documentation says to set "pipelining = True" to make that particular error go away but in my case the error started showing up after making it "true". With having pipelining set to False, it of course transfers a python script to z/OS and tries to run which fails for some other reason:

"The TSO command \"LU GPUSER\" execution failed.", "changed": false, "output": [{"command": "LU GPUSER", "rc": 16, "content": ["\udcc2\udcd7\udce7\udce6\udcf0\udcf0\udcf0\udcf0\udcc9@\u0167\udc85\udc83@\udc95\udc96\udca3@\udc86\udc96\udca4\udc95\udc84\u0015"], "lines": 1}], "failed": true, "invocation": {"module_args": {"commands": ["LU GPUSER"]}}}

Interestingly, if I intercept the temporary python script uploaded to the z/OS and manually run it via python interpreter on my z/OS unix shell it actually works and gives me good output. 

Does it have something to do with conversion (or lack of it) from ascii to ebdic? 
Please let me know if you know what I have missed. thanks!

**Comment by rexemin** [comment, marked as answer]
Hi, thanks for reaching out to the team, could you tell us which version of the collection you're using as well as which version of Ansible? It's worth noting we currently don't certify that our collection will work completely with ZOAU versions greater than 1.1.1 nor with Python 3.10, we're working on that for our next releases.
It could be that your issue is related to the environment variables you have set in all.yml and host_vars.yml. When you run the python script manually, are there any differences on the environment/path set on your shell and the one you configure in Ansible?
It would also help if you could run the playbook in verbose mode with `-vvv` and share the output it displays when running the task that is causing you problems.

**Reply by gpow81** [reply]
_Edit: I've added a couple of other findings at the bottom which probably are much more related to the actual issue. But I will leave bellow for the full picture (although I am pretty sure now that my problem is not related to ansible or the collections at all)._   

Thank you for trying to help! Interesting, if we can't find a smoking gun then I will try to downgrade these elements. 
I am attaching the output from the job with -vvv.
In one of the tasks I am printing out all the variables to verify that the script is using them and it looks like it does. Here it the actual yml file
 
```
- hosts: zos_host
  collections:
  - ibm.ibm_zos_core
  gather_facts: no
  tasks:
    - name: list all variables
      debug:
        var: hostvars[inventory_hostname]

    - name: Execute TSO command
      zos_tso_command:
          commands:
              - netstat
      register: result
    - debug:
        msg: "{{ result }}"
```
[out.log](https://github.com/ansible-collections/ibm_zos_core/files/9385744/out.log)

When I run the intercepted temporary python script manually on the z/OS, I don't think it uses any variables from the playbook but I might be wrong about it. I basically just run 
```
python AnsiballZ_zos_tso_command.py    (this is the script which I've intercepted when I had run the playbook with pipelining = False)

{"changed": true, "output": [{"command": "netstat", "rc": 0, "content": ["MVS TCP/IP NETSTAT CS V2R5       TCPIP Name: TCPIP           23:17:35", "User Id  Conn     State", "-------  ----     -----", "AZFSTC   0000003D Listen  ", "  Local Socket:   127.0.0.1..6790   ... [cut]                    
```               


BTW, I had a crazy idea to run this exact playbook against a x86 linux server (I needed to change the path for the python) and 
of course it didn't work but it got somewhat further, meaning it at least started reading the python program:

```
WARNING: The below traceback may *not* be related to the actual failure.
  File "/tmp/ansible_zos_tso_command_payload_c1zAZV/ansible_zos_tso_command_payload.zip/ansible/module_utils/basic.py", line 2592, in run_command
    cmd = subprocess.Popen(args, **kwargs)
  File "/usr/lib/python2.7/subprocess.py", line 394, in __init__
    errread, errwrite)
  File "/usr/lib/python2.7/subprocess.py", line 1047, in _execute_child
    raise child_exception
fatal: [zos_host]: FAILED! => {
    "changed": false,
    "cmd": "/tmp/tmpJBFDVB netstat",
    "invocation": {
        "module_args": {
            "commands": [
                "netstat"
            ]
        }
    },
    "msg": "[Errno 26] Text file busy",
    "rc": 26
}
```

This makes me think that my problem has nothing to do with zoau since it's not even get to the point of using any of it. 
I suspect the problem is with either python or ansible or something weird happens during pipeing the program over the ssh. 


edit:
narrowing it down I think,
bellow should work just as is right? (works if I do that against an x86 host)
```
ssh gpowied@zos1 /bin/sh -c '/usr/lpp/IBM/cyp/v3r10/pyz/bin/python'
print('Hello, world!')
SyntaxError: Non-UTF-8 code starting with '\x97' in file <stdin> on line 2, but no encoding declared; see https://python.org/dev/peps/pep-0263/ for details
```

Doesn't make a difference if I send # -*- coding: utf-8 -*-  as a firs line

So now my eyes are on the python itself ... I will look for a way to downgrade it but I am hopping for other clues. 


edit2: 
I've noticed that depending what I try to printf to python via SSH, the HEX code changes in the error. After few tries I think I can be sure that it treats these as EBCDIC but clearly it expects UTF-8.  So ssh is probably doing that. Is that a good clue? 
 ```
printf 'qrint("Hello, world")' | ssh gpowied@zos1 "PYTHONUTF8=1 /usr/lpp/IBM/cyp/v3r10/pyz/bin/python"
SyntaxError: Non-UTF-8 code starting with '\x98' in file <stdin> on line 2, but no encoding declared; see https://python.org/dev/peps/pep-0263/ for details
```
"qrint" was on purpose. The "q" EBCDIC is x98 in HEX

**Reply by gpow81** [reply]
Can I ask someone to run this simple test?
printf 'print("Hello, world")' | ssh user@zos "/bin/sh -c /path_to_python/bin/python"

**Reply by gpow81** [reply]
I've switched to python3.8 and now hello world works!  But, ansible is giving me exact same error as I had before with pipelining OFF. 
So now it looks like this:

```
Using module file /root/.ansible/collections/ansible_collections/ibm/ibm_zos_core/plugins/modules/zos_tso_command.py
Pipelining is enabled.
  3039 1661179177.82748: _low_level_execute_command(): starting
  3039 1661179177.82755: _low_level_execute_command(): executing: /bin/sh -c '/tmp/python/usr/lpp/IBM/cyp/v3r8/pyz/bin/python3 && sleep 0'
<zos1> ESTABLISH SSH CONNECTION FOR USER: gpowied
<zos1> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o 'User="gpowied"' -o ConnectTimeout=10 -o ControlPath=/root/.ansible/cp/6911bd2538 zos1 '/bin/sh -c '"'"'/tmp/python/usr/lpp/IBM/cyp/v3r8/pyz/bin/python3 && sleep 0'"'"''
  3039 1661179177.83109: Sending initial data
  3039 1661179177.91927: Sent initial data (117092 bytes)
  3039 1661179178.58019: stdout chunk (state=3):
>>>
{"msg": "The TSO command \"netstat\" execution failed.", "changed": false, "output": [{"command": "netstat", "rc": 16, "content": ["\udcc2\udcd7\udce7\udce6\udcf0\udcf0\udcf0\udcf0\udcc9@\u0167\udc85\udc83@\udc95\udc96\udca3@\udc86\udc96\udca4\udc95\udc84\u0015"], "lines": 1}], "failed": true, "invocation": {"module_args": {"commands": ["netstat"]}}}
<<<

```

**Reply by rexemin** [reply]
Thank you for the additional information, a couple of things to note: since you're using Ansible 2.9.6, you should keep using `pipelining = False`. Additionally, Ansible is not using the environment variables you configured in your inventory, in the output you shared the task is run like this: 
`SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o 'User="gpowied"' -o ConnectTimeout=10 -o ControlPath=/root/.ansible/cp/6911bd2538 zos1 '/bin/sh -c '"'"'/usr/lpp/IBM/cyp/v3r10/pyz/bin/python3.10 && sleep 0'"'"''`

When it should look like this: 
`SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o Port=22 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ‘User=“omvsadm”’ -o ConnectTimeout=10 -o ControlPath=/Users/iamorenosoto/.ansible/cp/de618e5fd4 -tt [ec01134a.vmec.svl.ibm.com](http://ec01134a.vmec.svl.ibm.com/) ‘/bin/sh -c ‘“’”‘_BPXK_AUTOCVT=ON ZOAU_HOME=/zoau/v1.2.0 PYTHONPATH=/zoau/v1.2.0/lib LIBPATH=/zoau/v1.2.0/lib:/python3/usr/lpp/IBM/cyp/v3r8/pyz/lib:/lib:/usr/lib:. PATH=/zoau/v1.2.0/bin:/python3/usr/lpp/IBM/cyp/v3r8/pyz/bin:/bin:/var/bin _CEE_RUNOPTS=‘“’”‘“’”‘“’”‘“’”‘FILETAG(AUTOCVT,AUTOTAG) POSIX(ON)‘“’”‘“’”‘“’”‘“’”' _TAG_REDIR_ERR=txt _TAG_REDIR_IN=txt _TAG_REDIR_OUT=txt LANG=C /python3/usr/lpp/IBM/cyp/v3r8/pyz/bin/python3.8 /.ansible/tmp/ansible-tmp-1660955267.8000598-43378-135487875126096/AnsiballZ_zos_mvs_raw.py && sleep 0'“‘”’'`

Are you sourcing the environment variables in your playbook? If you take a look at [this sample playbook](https://github.com/IBM/z_ansible_collections_samples/blob/master/zos_concepts/zos_operator/zos_operator_basics/zos_operator_basics.yml) you can source them with this line before the tasks: `environment: "{{ environment_vars }}"`

**Reply by gpow81** [reply]
This was it! At some point in my desperation I've dropped that line and I was sure that variables were still being sourced anyway because they were still showing up in my "list all variables" task . But I hadn't  noticed that they have stopped showing on the ssh line! 

pipelining = True  - works only with python 3.8  and not 3.10.5
pipelining = False  works with python 3.8 and 3.10.5 

Big thank you!

**Reply by ddimatos** [reply]
We are working with the Python team to see if they can support pipelining=True for Python 3.10 and later.

**Comment by ddimatos** [comment]
The latest PTF/patch from IBM Enterprise Python for 3.10 and 3.11 supports a new environment variable that allows for `pipelining=true` to operate as it did in the past. Tha environment var is `PYTHONSTDINENCODING: "cp1047"`. See git discussion #657 for an example.
